File: July-LeetCoding-Challenge-2020/Week2/430-flatten-a-multilevel-doubly-linked-list.js
/**
 * // Definition for a Node.
 * function Node(val,prev,next,child) {
 *    this.val = val;
 *    this.prev = prev;
 *    this.next = next;
 *    this.child = child;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 * Runtime: 64 ms, faster than 82.79% of JavaScript online submissions for Flatten a Multilevel Doubly Linked List.
 * Memory Usage: 37.5 MB, less than 5.33% of JavaScript online submissions for Flatten a Multilevel Doubly Linked List.
 */
var flatten = function(head) {
    if (!head) {
        return head;
    }

    let result = flattenHelper(head);

    result.prev = null;

    return result;
};

function flattenHelper (node) {
    if (!node) {
        return node;
    }

    const nodeCopy = node;
    while (node.next && !node.child) {
        node = node.next;
    }

    const child = flattenHelper(node.child);

    if (child) {
        child.prev.next = node.next;

        if (node.next) {
            node.next.prev = child.prev;
        }

        child.prev = node;
        node.next = child;
    }

    node.child = null;

    while (node.next) {
        node = node.next;
    }

    nodeCopy.prev = node;

    return nodeCopy;
}

File: July-LeetCoding-Challenge-2020/Week2/430-flatten-a-multilevel-doubly-linked-list.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Node') -> 'Node':
        # DFS 
        # Stack
        if not head:
            return head
        trace=Node(0, None,None,None)
        curr = trace
        stack = [head]
        while stack:
            stack_pop = stack.pop()
            if stack_pop.next:
                stack.append(stack_pop.next)
            if stack_pop.child:
                stack.append(stack_pop.child)
            curr.next = stack_pop
            stack_pop.prev = curr
            stack_pop.child = None
            curr = stack_pop
        trace.next.prev = None # head pointing trace
        return trace.next
        # TC: O(n), need to check every element
        # SC: O(n), worst case = all child, if we exclude stack, O(1)
        